### PR TITLE
Python: WIP for Command Injection query using new dataflow

### PR DESCRIPTION
**Pull request by RasmusWL**
This PR does a few things in one go. You might wonder why it wasn't split up, but the things kinda evolved together, and having a query that gives no results because there are no sources isn't much fun :smile: 

1. Port 'command injection' query to new dataflow (easy)
2. Add helpers for resolving imports
3. Model `os.system` and `os.popen` calls as `SystemCommandExecution::Range`
4. Model Flask requests

Part 4 was by FAR the most demanding, and I'm not sure we're at a top-notch polished solution yet. However, things have improved quite a bit from my frankensteins monster I had yesterday :japanese_ogre: 

On the way I experimented with how the path explanations would end up looking like when using TypeTracker (in combination with taint-tracking) for defining additional taint steps. I added some tests files so this can be replicated by others as well (https://github.com/github/codeql/commit/a82fa04d8a49fa8147d484a6798c6c397216d90e#diff-55b2593c15fcf2941d07ae0a613f4fe0). 

How to actually use them in a "clean" way was a bit surprising, but ultimately let me remove a lot of spaghetti code. The conclusion was that for defining adding additional taint steps, you _don't_ want to be using the result of a TypeTracker `t` that `t.startsIn("attr_name")`, instead you want to only add additional steps EXACTLY where the lookup is happening (so using `AttrNode`). HOWEVER, that means supporting something like `getattr` in TypeTracker will not end up giving a net-benefit, since the code is not handling this... I don't really know how to handle this in a good way, but maybe we need a `DataFlow::Node attrAccess(DataFlow::Node, string attr_name)` helper predicate that can encapsulate this. Any thoughts on this would be much appreciated.

What I was able to achieve with the Flask request modeling is that (1) "easy" attributes that are already supported just works (`str`, `List[str]`, and such), and (2) I experimented with modeling a custom class (MultiDict) that some of the request attributes are instances of and was able to define additional taint steps for the custom `getlist("key")` method. 

### Still not handled

1. Somehow we should be able to specify that the elements of `request.files` (a MultiDict) belong to the class `Werkzeug::Datastructures::FileStorage` and should have the additional taint steps defined there AND that the 0-indexed argument to its' [save method](https://werkzeug.palletsprojects.com/en/1.0.x/datastructures/#werkzeug.datastructures.FileStorage.save) is a sink for path-injection.

  We might be able to get _some_ of the way "for free" if we adopt the access-path based modeling of library functions that C# uses, but I'm not sure whether we can use that to define the set of `DataFlow::Node`s that (could) contain an instance of `werkzeug.datastructures.FileStorage` so we can define sinks and additional taint steps. (These are just ideas, I don't know how to phrase that into a question that the C# team can answer).

  JS might have a solution to this already, and I will try to ask them directly.

2. We currently don't model other special fields of `flask.request` that are not `MultiDict`s -- these have been marked by TODOs, and doesn't seem to pose any unique challenges. (so it's just grunt work).

### Annoyances

Whenever I need to track an attribute of an object, I seem to write this piece of code:

```ql
private DataFlow::Node some_attr_tracking(DataFlow::TypeTracker t) {
  t.startInAttr("<attr_name>") and
  result = <qualifier specification>
  or
  exists(DataFlow::TypeTracker t2 | result = some_attr_tracking(t2).track(t2, t))
}

DataFlow::Node some_attr_tracking() {
  result = some_attr_tracking(DataFlow::TypeTracker::end())
}
```

If I need to track two attributes separately, that requires two separate predicates... and that becomes A LOT of boilerplate code. 

I'm kinda envious when looking at the JS modeling code (we need to compare to the [pre-API-graph code](https://github.com/github/codeql/pull/4082/files#diff-9aa94c4d713ef9d8da73918ff53db774L33)) where they have fancy helper predicates such as [`getAMemberCall(string memberName)`](https://github.com/github/codeql/blob/8f68f512dfdf62efd67e119c0d94017183827384/javascript/ql/src/semmle/javascript/dataflow/Sources.qll#L104-L111) that just handles all of the bound-method boilerplate work out of the box.

I *do* understand that it probably won't be as easy as defining a predicate like

```ql
DataFlow:Node trackedAttr(DataFlow::Node qualifier, string name, DataFlow::TypeTracker t) {
  t.startInAttr(name) and
  result = qualifier
  or
  exists(DataFlow::TypeTracker t2 | result = trackedAttr(qualifier, name, t2).track(t2, t))
}
```

since that will contain ALL attribute accesses (and what they can be tracked to) for ALL objects. @tausbn do you see a clever solution to this? :blush: 

---

PS. The name `RequestInputAccess` was borrowed from [JS](https://github.com/github/codeql/blob/32b0f1b480a279573687ac277023b727a6c657f6/javascript/ql/src/semmle/javascript/frameworks/HTTP.qll#L479).

### CODE REVIEW

File: python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll [hunk 1]
```diff
@@ -7,3 +7,70 @@ private import python
...
+      exists(DataFlow::TypeTracker t2 | result = os::system(t2).track(t2, t))
+    }
+
+    /** Gets a reference to the `os.system` function. */
+    DataFlow::Node system() { result = os::system(DataFlow::TypeTracker::end()) }
+
+    /** Gets a reference to the `os.popen` function. */
+    DataFlow::Node popen(DataFlow::TypeTracker t) {
+      t.start() and
+      result = DataFlow::importMember("os", "system")
```
Comment (yoff) [inline python/ql/src/experimental/semmle/python/frameworks/Stdlib.qll:41]:
```suggestion
      result = DataFlow::importMember("os", "popen")
```

File: python/ql/test/experimental/library-tests/frameworks/modeling-example/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+We want to extend our analysis to `obj.get_value()` is also tainted if `obj` is a tainted instance of `MyClass`.
+
+The actual type-tracking is done in `SharedCode.qll`, but it's the _way_ we use it that matters.
+
+In `NaiveModel.ql` we add an additional taint step from an instance of `MyClass` to calls of the bound method `get_value` (that we have tracked). It provides us with the correct results, but the path explanations are not very useful, since we are now able to cross functions in _one step_.
+
+In `ProperModel.ql` we split the additional taint step in two:
+
+1. from tracked `obj` that is instance of `MyClass`, to `obj.get_value` **but only** exactly where the attribute is accessed (by an `AttrNode`). This is important, since if we allowed `<any tracked qualifier>.get_value` we would again be able to cross functions in one step.
+2. from tracked `get_value` bound method to calls of it, **but only** exactly where the call is (by an `CallNode`). for same reason as above.
```
Comment (yoff) [inline python/ql/test/experimental/library-tests/frameworks/modeling-example/README.md:23]:
```suggestion
2. from tracked `get_value` bound method to calls of it, **but only** exactly where the call is (by a `CallNode`). for same reason as above.
```

**Comment by RasmusWL** [comment]
From talking to JS team, they use some pseudo properties that can be used to track element accesses.

Some of the interesting collection definitions are in https://github.com/github/codeql/blob/f618d430e7274de8ddd5c0140e7009f9856a2668/javascript/ql/src/semmle/javascript/Collections.qll#L16-L25

and example usage in

https://github.com/github/codeql/blob/3f8881a334396fe8c6b11cd65f09b156b307ca1d/javascript/ql/src/Security/CWE-327/BadRandomness.ql#L58-L70

**Comment by RasmusWL** [comment]
GitHub action tests are stuck in pending, but other tests are green now :tada: 

Even though this is still a bit WIP, I think we can merge this PR now, and subsequent PRs to improve modeling will probably be a bit easier to review :wink:

**Review by tausbn — APPROVED** [review]
In the spirit of "get it merged and then make it nicer later on" I only have one small comment. 
This looks ready to merge to me.

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll [hunk 3]
```diff
@@ -458,7 +462,9 @@ predicate listStoreStep(CfgNode nodeFrom, ListElementContent c, CfgNode nodeTo)
   //   nodeFrom is `42`, cfg node
   //   nodeTo is the list, `[..., 42, ...]`, cfg node
   //   c denotes element of list
-  nodeTo.getNode().(ListNode).getAnElement() = nodeFrom.getNode()
+  nodeTo.getNode().(ListNode).getAnElement() = nodeFrom.getNode() and
+  // Suppress unused variable warning
+  c = c
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll:467]: @yoff :point_up:

**Comment by RasmusWL** [comment]
:shipit:

**Pull request merged**